R/BASS_studyArea_analysis.R

```diff
@@ -1,96 +1,224 @@
-      #  n_os <- lapply(FUN = function(x) x * os, X = nARUs )
-      #  if(n_os==0) n_os <- NULL
-      #} else if ( all(strata_vector %in% names(nARUs)) ){
-      n_os <- os
-
-      if(!length(os) %in% c(1, length(nARUs))) {
-        rlang::abort(
-          c("x" = "Not all strata found in `os`, but `os` has length > 1",
-            "*" = paste0("`os` should either be single value or named vector, ",
-                         "one for each stratum.")))
+    # Get n_strata and n_os depending on inputs
+
+    # Check data frame
+    if(inherits(nARUs, "data.frame")) {
+
+      # Problem: Miss named or wrong strata
+      if(!all(names(nARUs) %in% c(stratum_name, "n", "n_os")) ||
+         !all(nARUs[[stratum_name]] %in% strata_vector)) {
+        abort_strat()
+      }
+
+      # Convert from data frame
+      n_os <- as.list(nARUs$n_os) |>
+        stats::setNames(nARUs[[stratum_name]])
+      n_strata <- as.list(nARUs$n) |>
+        stats::setNames(nARUs[[stratum_name]])
+
+    } else {
+      # Check list (convert if vector)
+      nARUs <- as.list(nARUs)
+      if(length(os) > 1) os <- as.list(os)
+
+      # Problem: List not named correctly
+      if(!(rlang::is_named(nARUs) && all(names(nARUs) %in% strata_vector))) {
+        abort_strat()
+      }
+
+      # Problem: List not named correctly (and not length = 1)
+      if(!((rlang::is_named(os) && all(names(os) %in% strata_vector)) ||
+           length(os) == 1)) {
+        abort_strat("`os` must be a single value, or a vector/list named by strata")
       }
-    }  else {
-      rlang::abort(
-        c("x" = "Not all strata found in `nARUs`, but `nARUs` has length >1",
-          "*" = paste0("`nARUs` should either be single value or named vector, ",
-                       "one for each stratum.")))
+
+      n_strata <- nARUs
+
+      if(!rlang::is_named(os) && length(os) == 1) {
+        n_os <- lapply(n_strata, \(x) round(x * os))
+
+        # If all 0, use NULL
+        if(all(vapply(n_os, \(x) x == 0, logical(1)))) n_os <- NULL
+      } else n_os <- os
+    }
+
+    # Problem: Chose stratification, but only one strata
+    if(length(n_strata) == 1 || (rlang::is_named(n_os) && length(n_os) == 1)) {
+      abort_strat("There is only one stratum")
     }
   }
 
-  purrr::map(
-    1:num_runs,
-    ~ spsurvey::grts(sframe = attframe,
-                     n_over = n_os,
-                     n_base = Stratdsgn,
-                     stratum_var = stratum_name,
-                     mindis = mindis,
-                     DesignID = "sample",
-                     aux_var = "inclpr",
-                     maxtry = maxtry)
-  )
+  s <- set_seed(seed, {
+    purrr::map(
+      1:num_runs,
+      # Must be \(x) to use ... (otherwise ... overwritten)
+      \(x) spsurvey::grts(sframe = probs,
+                          n_over = n_os,
+                          n_base = n_strata,
+                          stratum_var = stratum_name,
+                          DesignID = "sample",
```

Comment:
we will want to push DesignID to the `...`. I think for `aux_var` we want users to generate and specify their own columns to use here. That doesn't seem possible with the `...`, so perhaps need to include another variable with a default of "inclpr".

Reply:
I'd just want to make sure doing so won't break other places in the code if "inclpr" is needed to be the selection column.